### ISSUE TITLE
Statistics collection and display

### ISSUE THREAD

**Issue by markmandel** [googleforgames/agones#144]
Stats around game server performance, as well as in Game statistics need to be easily exportable from the GameServer into Agon through the SDK and go to some kind of statistic data store (such as [Prometheus](https://prometheus.io) either directly, or through [OpenCensus](https://opencensus.io)). There will also need to be some kind of visualisation layer, and possibly alerting as well.

Discussion will need to cover (a) what types of stats they would want to capture (b) the SDK API to collect them, and (c) the visualisation and/or alerting requirements

# Statistics Ideas
- Number of game servers / ready / allocated / unhealthy / etc
- Game server CPU / Memory - aggregate and per game server
- Game server Frame rates
- Game server hitching
- Scale up time of nodes, and corresponding node resource sizes (we'll need this for node autoscaling)
- Visualization of game servers/nodes/vs current trend up/down scaling rates

# Research
- https://github.com/improbable-eng/thanos

**Comment by kwiesmueller** [comment]
Wouldn't it be best to really use Prometheus directly and then us the included service discovery to ensure metrics from all nodes get scraped?
I can't really tell a reason why there should even be some way to send metrics through agones as it would only implement more complexity than necessary.
The only thing I could imagine is the need for access to the metrics from inside some components (including game servers) which could be solved as well.

For instance, there would already be a quite good integration with GCP and Stackdriver as well and as far as I know Prometheus is quite widely adopted too.

**Comment by markmandel** [comment]
Coming back to this - there are some very widely used statistics that we should provide out of the box - it seems like duplication of work to have everyone do this themselves. We will also need some standard visualisations of what is happening inside Agones as well.

That being said (and one of the reasons I like OpenCensus) - if there is a way we can expose/leverage a project like that to allow stats to go anywhere without blocking stats collection.

Basically find a way to make the best of both worlds.

**Comment by GabeBigBoxVR** [comment]
Wow, this would be amazing, especially for FPS games where tick rate is super important.

**Comment by Kuqd** [comment]
I'm going to start things off very simple using opencensus and prometheus in the controller, then we can build from there and add more metrics.
I'll start with : Number of game servers / ready / allocated / unhealthy

**Comment by markmandel** [comment]
@heartrobotninja just checking in - did you have any work here that @Kuqd could leverage?

**Comment by Kuqd** [comment]
I'm planning to collect also:
- go metrics
- healtheck metrics
- process metrics

**Comment by Kuqd** [comment]
Hey Guys,

I've almost finished my work, but I ran into some issues trying to add total of gameservers and allocations metrics.
State metrics are working very well using informers cache.

# First why totals ? (may be this not even interesting)

Total counter as opposed to current gauge would allow user to aggregate using rates of gameserver per status (errors per sec, allocated per sec, etc), the same goes for allocations.
In prometheus you can only rate or irate over count that only increment.

# Problems

## Using informers cache 

I tried using informers (see https://github.com/Kuqd/agones/blob/features/opencensus/pkg/metrics/controller.go#L260) for final states like `ERROR` or `ALLOCATED` or `READY` it works pretty well, however for transition state like `CREATING` if those transition are too fast, I'm not going to catch an event within the informers (those event are collapsed and this is by design see [documentation](https://github.com/kubernetes/client-go/blob/master/tools/cache/shared_informer.go#L34) )

## Using watch

I've tried another implementation using a [watch for fleet allocations](https://github.com/Kuqd/agones/blob/features/opencensus/pkg/metrics/controller.go#L279) but if the controller crashes or stops the watches start from the begining which means multiple scrape will count the data twice. I've been thinking to use annotations on the controller to remember the last `ressourceVersion` notification I've received to give as argument for the watch, but that seems complicated.

# Conclusion

May be the whole solution is to add those total metrics within Agones code, but I wanted to avoid to spread metrics code across the project.

Lastly I've not been able to remove metrics for past state resource using OpenCensus `LastValue()` aggregation so far my solution is to [zero data](https://github.com/Kuqd/agones/blob/features/opencensus/pkg/metrics/gameservers.go#L36) on deletion but may be there is .a better way ? The problem with this is that the amount of tim series will grow over time even after resource deletion, more importantly for fleets.



What do you think ? I'll appreciate feedback, I'm thinking to add a func in the metrics package to be used by other controller.

**Comment by markmandel** [comment]
> May be the whole solution is to add those total metrics within Agones code, but I wanted to avoid to spread metrics code across the project.

I assume that this would do the count at time of processing? 

So count `Starting` somewhere around here:
https://github.com/GoogleCloudPlatform/agones/blob/master/pkg/gameservers/controller.go#L418

If we did this though, we would still skip transition states, because that's what we'll be processing anyway.

So it seems like the informers cache is most accurate - as it's actually following what is being processed.

Am I off base here? It feels like that is the winner?

I think in some places, we're going to have to put metrics inside controllers - but this likely scales better for each resource type, code wise. Rather than interleaving metrics code throughout each controller, you can list out the resources you want to listen to, and add event handlers for all of them in one go - which makes sense to me.

**Comment by Kuqd** [comment]
Yes make sense, however I can’t use the add handler for totals as it’s currently getting triggered by the resync. For gameserver it’s fine, I’m just catching changes, for fleet allocation I need to run some tests.

**Comment by markmandel** [comment]
The biggest issue I see with watch + annotations - is that it's _yet another_ write to etcd that we are doing over and over. So I have some concern there?

Unless maybe we tie it into status updates? I'm not sure. 

As I said in chat, it may be worth asking people in Slack #sig-apimachinery or #kubernetes-operators to see how they are doing things. 

It may also be at a point where it's better to start with _something_ and then we can iterate if we decide it's not the right approach once we have something to look at.

**Comment by Kuqd** [comment]
We should add metrics for ping service.

**Comment by Kuqd** [comment]
slack discussion: we should instrument go-client with opencensus

**Comment by markmandel** [comment]
@Kuqd I had a memory of us talking about closing this ticket. Are we at a place where we want to do so, or not yet?

**Comment by markmandel** [comment]
Gentle bump on this @Kuqd  - shall we close this, and have a new ticket(s) for extra metrics (cpu tick rate, player counts, etc)

**Comment by Kuqd** [comment]
Do we have metrics for the new allocation? Sorry been off lately

**Comment by markmandel** [comment]
@Kuqd Ah, good point - no we don't. I was splitting up #600 and fell behind on it a little, so `GameServerAllocation` is not yet switched to an API Endpoint. Current PR in flux in #659
